File: cmd/leaderboard/workloads.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"strings"

	"gopkg.in/yaml.v3"

	"github.com/inference-sim/leaderboard/internal/catalog"
	"github.com/inference-sim/leaderboard/internal/schema"
)

// The workload catalog endpoints (design §5). They read and write workloads.yaml, the
// one (local) file of named workload profiles, and validate a proposed profile —
// handing a raw WorkloadSpec to blis itself so the errors are blis's own, not a port.
//
// profileBody is the wire shape of one profile, shared by GET output and
// create/edit/validate input. It mirrors the clean catalog YAML shape: a distribution
// profile carries the flat fields, a spec profile carries `spec` (+ its hash on
// output). Kept separate from catalog.Profile so the HTTP contract is explicit.
type profileBody struct {
	Name            string       `json:"name"`
	Seed            int64        `json:"seed"`
	HorizonTicks    *int64       `json:"horizon_ticks"`
	RequestTimeoutS *int         `json:"request_timeout_s"`
	Workload        workloadBody `json:"workload"`
	// Builtin is output-only: true for a shipped preset, so the catalog browser badges it
	// and disables Edit/Rename/Delete (§11.2). It is ignored on input — a request cannot
	// declare itself a preset.
	Builtin bool `json:"builtin,omitempty"`
}

type workloadBody struct {
	Type              string    `json:"type"`
	NumRequests       int       `json:"num_requests,omitempty"`
	Load              *loadBody `json:"load,omitempty"`
	PromptTokens      int       `json:"prompt_tokens,omitempty"`
	PromptTokensStdev int       `json:"prompt_tokens_stdev,omitempty"`
	OutputTokens      int       `json:"output_tokens,omitempty"`
	OutputTokensStdev int       `json:"output_tokens_stdev,omitempty"`
	SpecSHA256        *string   `json:"spec_sha256,omitempty"`
	// Spec is the inline WorkloadSpec as an object; SpecYAML is the same authored as
	// YAML text. The editor sends SpecYAML (the browser has no YAML parser and the raw
	// mode is a text area); the server parses it here with the same parser blis uses.
	// GET returns both, so the editor can round-trip the text. SpecYAML wins if both
	// are present.
	Spec     map[string]any `json:"spec,omitempty"`
	SpecYAML *string        `json:"spec_yaml,omitempty"`
}

type loadBody struct {
	Kind  string  `json:"kind"`
	Value float64 `json:"value"`
}

// validateResponse is what /api/workloads/validate returns: whether the profile is
// runnable, the variant it would be stored as, the one-line work summary, the name of
// a content twin if any (P3), and every reason it is not yet runnable.
type validateResponse struct {
	OK      bool     `json:"ok"`
	Variant string   `json:"variant"`
	Summary string   `json:"summary"`
	Twin    *string  `json:"twin"`
	Issues  []string `json:"issues"`
}

// toProfile resolves a wire body into a catalog profile, applying the request-timeout
// default the same way the file loader does and parsing a raw spec_yaml if the client
// sent one instead of a spec object.
func (b profileBody) toProfile() (catalog.Profile, error) {
	timeout := 300
	if b.RequestTimeoutS != nil {
		timeout = *b.RequestTimeoutS
	}
	w := catalog.Workload{
		Type:              b.Workload.Type,
		NumRequests:       b.Workload.NumRequests,
		PromptTokens:      b.Workload.PromptTokens,
		PromptTokensStdev: b.Workload.PromptTokensStdev,
		OutputTokens:      b.Workload.OutputTokens,
		OutputTokensStdev: b.Workload.OutputTokensStdev,
		Spec:              b.Workload.Spec,
	}
	if b.Workload.SpecYAML != nil && strings.TrimSpace(*b.Workload.SpecYAML) != "" {
		var spec map[string]any
		if err := yaml.Unmarshal([]byte(*b.Workload.SpecYAML), &spec); err != nil {
			return catalog.Profile{}, fmt.Errorf("the spec is not valid YAML: %v", err)
		}
		w.Spec = spec
	}
	if b.Workload.Load != nil {
		w.Load = schema.Load{Kind: b.Workload.Load.Kind, Value: b.Workload.Load.Value}
	}
	return catalog.Profile{
		Name:            b.Name,
		Seed:            b.Seed,
		HorizonTicks:    b.HorizonTicks,
		RequestTimeoutS: timeout,
		Workload:        w,
	}, nil
}

// profileToBody is the reverse, for GET output. A spec profile carries its content
// hash so the client can show it without recomputing.
func profileToBody(p catalog.Profile) profileBody {
	timeout := p.RequestTimeoutS
	b := profileBody{
		Name:            p.Name,
		Seed:            p.Seed,
		HorizonTicks:    p.HorizonTicks,
		RequestTimeoutS: &timeout,
		Builtin:         p.Builtin,
		Workload:        workloadBody{Type: p.Workload.Type},
	}
	switch p.Workload.Type {
	case "distribution":
		b.Workload.NumRequests = p.Workload.NumRequests
		b.Workload.Load = &loadBody{Kind: p.Workload.Load.Kind, Value: p.Workload.Load.Value}
		b.Workload.PromptTokens = p.Workload.PromptTokens
		b.Workload.PromptTokensStdev = p.Workload.PromptTokensStdev
		b.Workload.OutputTokens = p.Workload.OutputTokens
		b.Workload.OutputTokensStdev = p.Workload.OutputTokensStdev
	case "workload-spec":
		if sha, err := schema.SpecSHA256(p.Workload.Spec); err == nil {
			b.Workload.SpecSHA256 = &sha
		}
		b.Workload.Spec = p.Workload.Spec
		if y, err := yaml.Marshal(p.Workload.Spec); err == nil {
			s := string(y)
			b.Workload.SpecYAML = &s
		}
	}
	return b
}

// registerWorkloadRoutes adds the catalog endpoints to a mux. Go's method-and-path
// patterns pick the literal /validate over the {name} wildcard, so the two never
// collide, and they coexist with the older method-less /api/run pattern.
func (s *server) registerWorkloadRoutes(mux *http.ServeMux) {
	mux.HandleFunc("GET /api/workloads", s.handleWorkloadsList)
	mux.HandleFunc("POST /api/workloads", s.handleWorkloadCreate)
	mux.HandleFunc("POST /api/workloads/validate", s.handleWorkloadValidate)
	mux.HandleFunc("PUT /api/workloads/{name}", s.handleWorkloadUpdate)
	mux.HandleFunc("DELETE /api/workloads/{name}", s.handleWorkloadDelete)
}

// workloadsMux is a standalone mux carrying only the catalog routes, for tests.
func (s *server) workloadsMux() *http.ServeMux {
	mux := http.NewServeMux()
	s.registerWorkloadRoutes(mux)
	return mux
}

func (s *server) handleWorkloadsList(w http.ResponseWriter, r *http.Request) {
	c, err := catalog.Load(s.catalogPath)
	if err != nil {
		httpError(w, http.StatusInternalServerError, err.Error())
		return
	}
	bodies := make([]profileBody, 0, len(c.Profiles))
	for _, p := range c.Profiles {
		bodies = append(bodies, profileToBody(p))
	}
	writeJSON(w, http.StatusOK, map[string]any{"schema_version": catalog.SchemaVersion, "workloads": bodies})
}

// handleWorkloadValidate is the dry run: it never writes. It reports the variant, the
// work summary, a content twin if any, and the reasons the profile is not runnable —
// including blis's own errors for a raw spec.
func (s *server) handleWorkloadValidate(w http.ResponseWriter, r *http.Request) {
	body, err := decodeProfile(r)
	if err != nil {
		httpError(w, http.StatusBadRequest, err.Error())
		return
	}
	p, err := body.toProfile()
	if err != nil {
		httpError(w, http.StatusBadRequest, err.Error())
		return
	}
	c, err := catalog.Load(s.catalogPath)
	if err != nil {
		httpError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, s.validateProfile(p, c))
}

func (s *server) handleWorkloadCreate(w http.ResponseWriter, r *http.Request) {
	body, err := decodeProfile(r)
	if err != nil {
		httpError(w, http.StatusBadRequest, err.Error())
		return
	}
	s.catMu.Lock()
	defer s.catMu.Unlock()

	c, err := catalog.Load(s.catalogPath)
	if err != nil {
		httpError(w, http.StatusInternalServerError, err.Error())
		return
	}
	p, err := body.toProfile()
	if err != nil {
		httpError(w, http.StatusBadRequest, err.Error())
		return
	}
	if _, exists := c.ByName(p.Name); exists {
		httpError(w, http.StatusConflict, fmt.Sprintf("a workload named %q already exists; "+
			"edit it or choose another name", p.Name))
		return
	}
	if status, msg := s.rejectIfUnrunnableOrTwin(p, c); status != 0 {
		httpError(w, status, msg)
		return
	}
	c.Profiles = append(c.Profiles, p)
	if err := c.Write(s.catalogPath); err != nil {
		httpError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, profileToBody(p))
}

func (s *server) handleWorkloadUpdate(w http.ResponseWriter, r *http.Request) {
	target := r.PathValue("name")
	body, err := decodeProfile(r)
	if err != nil {
		httpError(w, http.StatusBadRequest, err.Error())
		return
	}
	s.catMu.Lock()
	defer s.catMu.Unlock()

	c, err := catalog.Load(s.catalogPath)
	if err != nil {
		httpError(w, http.StatusInternalServerError, err.Error())
		return
	}
	idx := -1
	for i, p := range c.Profiles {
		if p.Name == target {
			idx = i
			break
		}
	}
	if idx < 0 {
		httpError(w, http.StatusNotFound, fmt.Sprintf("no workload named %q to edit", target))
		return
	}
	if c.Profiles[idx].Builtin {
		httpError(w, http.StatusForbidden, fmt.Sprintf("%q is a built-in preset and is read-only; "+
			"author a new workload from it instead", target))
		return
	}
	p, err := body.toProfile()
	if err != nil {
		httpError(w, http.StatusBadRequest, err.Error())
		return
	}
	// A rename must not collide with a different existing profile.
	if p.Name != target {
		if _, exists := c.ByName(p.Name); exists {
			httpError(w, http.StatusConflict, fmt.Sprintf("a workload named %q already exists", p.Name))
			return
		}
	}
	// Twin detection excludes the profile being replaced: editing it in place, or
	// renaming it, is not a collision with its former self.
	others := &catalog.Catalog{}
	for i, existing := range c.Profiles {
		if i != idx {
			others.Profiles = append(others.Profiles, existing)
		}
	}
	if status, msg := s.rejectIfUnrunnableOrTwin(p, others); status != 0 {
		httpError(w, status, msg)
		return
	}
	c.Profiles[idx] = p
	if err := c.Write(s.catalogPath); err != nil {
		httpError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, profileToBody(p))
}

func (s *server) handleWorkloadDelete(w http.ResponseWriter, r *http.Request) {
	target := r.PathValue("name")
	s.catMu.Lock()
	defer s.catMu.Unlock()

	c, err := catalog.Load(s.catalogPath)
	if err != nil {
		httpError(w, http.StatusInternalServerError, err.Error())
		return
	}
	kept := make([]catalog.Profile, 0, len(c.Profiles))
	found := false
	for _, p := range c.Profiles {
		if p.Name == target {
			if p.Builtin {
				httpError(w, http.StatusForbidden, fmt.Sprintf("%q is a built-in preset and cannot "+
					"be deleted", target))
				return
			}
			found = true
			continue
		}
		kept = append(kept, p)
	}
	if !found {
		httpError(w, http.StatusNotFound, fmt.Sprintf("no workload named %q to delete", target))
		return
	}
	c.Profiles = kept
	if err := c.Write(s.catalogPath); err != nil {
		httpError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]string{"deleted": target})
}

// validateProfile runs every check without writing: our own well-formedness and P6
// model-pinning rules first (fast, and they must pass before blis is worth running),
// then — for a well-formed raw spec — blis itself. The twin is reported whether or not
// the profile is otherwise runnable, so the editor can warn early.
func (s *server) validateProfile(p catalog.Profile, c *catalog.Catalog) validateResponse {
	resp := validateResponse{Variant: p.Workload.Type, Summary: summarize(p), Issues: []string{}}
	if twin, ok := c.FindContentTwin(p); ok {
		t := twin
		resp.Twin = &t
	}
	if err := catalog.Validate(p); err != nil {
		resp.Issues = append(resp.Issues, err.Error())
		return resp
	}
	if p.Workload.Type == "workload-spec" && s.validateSpec != nil {
		if err := s.validateSpec(p.Workload.Spec); err != nil {
			resp.Issues = append(resp.Issues, err.Error())
			return resp
		}
	}
	resp.OK = len(resp.Issues) == 0
	return resp
}

// rejectIfUnrunnableOrTwin maps a create/update to an HTTP error: 422 when the profile
// is not runnable (our rules or blis), 409 when its content twins an existing profile
// (P3). A zero status means it may be written.
func (s *server) rejectIfUnrunnableOrTwin(p catalog.Profile, c *catalog.Catalog) (int, string) {
	v := s.validateProfile(p, c)
	if !v.OK {
		return http.StatusUnprocessableEntity, strings.Join(v.Issues, "; ")
	}
	if v.Twin != nil {
		return http.StatusConflict, fmt.Sprintf("this workload has identical content to %q; "+
			"a workload has one name (P3). Reuse it, or change a field", *v.Twin)
	}
	return 0, ""
}

func decodeProfile(r *http.Request) (profileBody, error) {
	var b profileBody
	dec := json.NewDecoder(r.Body)
	dec.DisallowUnknownFields()
	if err := dec.Decode(&b); err != nil {
		return profileBody{}, fmt.Errorf("could not read the profile: %v", err)
	}
	return b, nil
}

// summarize is the one-line work summary the validate response carries. It mirrors the
// web's workloadTitle phrasing for the distribution case; a spec is described by its
// aggregate rate when it declares one.
func summarize(p catalog.Profile) string {
	w := p.Workload
	if w.Type == "workload-spec" {
		if rate, ok := numberField(w.Spec["aggregate_rate"]); ok && rate > 0 {
			return fmt.Sprintf("spec-backed workload at %s req/s aggregate", trimFloat(rate))
		}
		return "spec-backed workload"
	}
	load := trimFloat(w.Load.Value) + " req/s"
	if w.Load.Kind == "concurrency" {
		load = trimFloat(w.Load.Value) + " concurrent sessions"
	}
	return fmt.Sprintf("%d requests at %s", w.NumRequests, load)
}

func numberField(v any) (float64, bool) {
	switch n := v.(type) {
	case float64:
		return n, true
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	case json.Number:
		f, err := n.Float64()
		return f, err == nil
	}
	return 0, false
}

func trimFloat(f float64) string {
	return strconv.FormatFloat(f, 'f', -1, 64)
}

File: web/src/workloads.ts
/**
 * The Workloads tab's logic, kept out of the components so every claim it makes is
 * testable — the same split as newrun.ts.
 *
 * A workload profile is the work offered minus model (the topology spec's workloadKey),
 * with a name. The editor authors one; `leaderboard serve` persists it to workloads.yaml
 * and validates it — handing a raw WorkloadSpec to blis itself, so the browser never
 * needs a YAML parser or blis's rules. The variant a profile is stored as is mechanical,
 * not a choice: the guided form is a distribution, the raw-spec mode is spec-backed.
 */

import type { RunRecord, WorkloadGroup } from './load'
import { workloadKey, workloadTitle } from './load'
import { defaultSpec, parseSpec, serializeSpec, setPath, summarizeSpec, type SpecObject } from './spec'

type Group = RunRecord['group']
type SchemaWorkload = Group['workload']

/** The wire shape of one profile — mirrors the server's profileBody (cmd/leaderboard). */
export interface WorkloadBody {
  type: 'distribution' | 'workload-spec'
  num_requests?: number
  load?: { kind: 'rate' | 'concurrency'; value: number }
  prompt_tokens?: number
  prompt_tokens_stdev?: number
  output_tokens?: number
  output_tokens_stdev?: number
  spec_sha256?: string | null
  /** The raw WorkloadSpec as YAML text; the server parses it. */
  spec_yaml?: string
  spec?: Record<string, unknown>
}

export interface ProfileBody {
  name: string
  seed: number
  horizon_ticks: number | null
  request_timeout_s: number
  workload: WorkloadBody
  /** True for a shipped preset (§11): read-only in the catalog. Output-only; the server
   * ignores it on input. */
  builtin?: boolean
}

/** The server's validate verdict: runnable?, variant, one-line summary, twin, reasons. */
export interface ValidateResponse {
  ok: boolean
  variant: string
  summary: string
  twin: string | null
  issues: string[]
}

/**
 * The editor's form state. The workload itself is authored as a WorkloadSpec, held as
 * the YAML buffer `specYaml` — the single carrier the form and the YAML pane both edit.
 * name/seed/horizon/timeout are the profile's group-side knobs, kept beside the spec.
 */
export interface FormValues {
  name: string
  seed: number
  horizonTicks: number | null
  requestTimeoutS: number
  specYaml: string
}

export interface Issue {
  field: string
  message: string
}

export interface Interpretation {
  issues: Issue[]
  body: ProfileBody | null
}

/** Mirrors internal/catalog.workloadNamePattern: lower-case letters, digits, dash,
 * underscore or dot, and non-empty. */
export const NAME_PATTERN = /^[a-z0-9._-]+$/

/** The form a New workload opens with: blis's own single-client spec defaults, so it is
 * runnable as authored. */
export function initialForm(): FormValues {
  return {
    name: '',
    seed: 42,
    horizonTicks: null,
    requestTimeoutS: 300,
    specYaml: serializeSpec(defaultSpec()),
  }
}

/** bodyToForm fills the editor from a saved profile, so Edit opens on its real values.
 * A spec profile opens on its YAML; a legacy distribution profile is lifted into an
 * equivalent WorkloadSpec so it too can be edited in the one editor. */
export function bodyToForm(body: ProfileBody): FormValues {
  const base = initialForm()
  const w = body.workload
  const specYaml =
    w.type === 'workload-spec'
      ? w.spec_yaml ?? (w.spec ? serializeSpec(w.spec as SpecObject) : '')
      : serializeSpec(distributionToSpec(w))
  return {
    ...base,
    name: body.name,
    seed: body.seed,
    horizonTicks: body.horizon_ticks,
    requestTimeoutS: body.request_timeout_s,
    specYaml,
  }
}

/** One token distribution's gaussian statistics, as the custom card and the converter
 * both express them: a mean and spread with a clamp. */
export interface TokenStat {
  mean: number
  std_dev: number
  min: number
  max: number
}

/** deriveMax is the clamp ceiling used when a mean/stdev pair carries no explicit max:
 * four standard deviations above the mean covers all but a negligible tail, so the
 * gaussian is not silently truncated. Shared so the card's default and the legacy
 * converter agree. */
export function deriveMax(mean: number, stdev: number): number {
  return Math.max(1, Math.round(mean + 4 * stdev))
}

/** gaussianSpec builds a single-client WorkloadSpec with gaussian input and output
 * distributions from explicit token statistics. Both the custom card (explicit min/max
 * from the user) and distributionToSpec (bounds derived for a legacy distribution) build
 * on it, so the two produce the same shape and the custom card is simply a guided front
 * end for a one-client spec. */
export function gaussianSpec(opts: {
  numRequests: number
  load: { kind: 'rate' | 'concurrency'; value: number }
  input: TokenStat
  output: TokenStat
}): SpecObject {
  const dist = (s: TokenStat) => ({
    type: 'gaussian',
    params: { mean: s.mean, std_dev: s.std_dev, min: s.min, max: s.max },
  })
  const client: SpecObject = {
    id: 'c0',
    arrival: { process: 'poisson' },
    input_distribution: dist(opts.input),
    output_distribution: dist(opts.output),
  }
  let spec: SpecObject = { version: '2', num_requests: opts.numRequests, clients: [client] }
  if (opts.load.kind === 'concurrency') {
    spec = setPath(spec, ['clients', 0, 'concurrency'], opts.load.value)
  } else {
    spec = setPath(spec, ['aggregate_rate'], opts.load.value)
    spec = setPath(spec, ['clients', 0, 'rate_fraction'], 1)
  }
  return spec
}

/** distributionToSpec lifts a legacy distribution workload into the equivalent
 * WorkloadSpec, so an existing profile opens in the spec editor rather than breaking. The
 * mean/stdev pair maps onto a blis gaussian, and the bounds are derived from the mean and
 * spread so the result validates. */
function distributionToSpec(w: WorkloadBody): SpecObject {
  const pMean = w.prompt_tokens ?? 0
  const pStd = w.prompt_tokens_stdev ?? 0
  const oMean = w.output_tokens ?? 0
  const oStd = w.output_tokens_stdev ?? 0
  return gaussianSpec({
    numRequests: w.num_requests ?? 0,
    load: w.load ?? { kind: 'rate', value: 0 },
    input: { mean: pMean, std_dev: pStd, min: 1, max: deriveMax(pMean, pStd) },
    output: { mean: oMean, std_dev: oStd, min: 1, max: deriveMax(oMean, oStd) },
  })
}

/** The variant every profile authored here is stored as: the editor is spec-first, so a
 * save is always a WorkloadSpec (the distribution variant is legacy, read on Edit only). */
export function variantOf(_v: FormValues): 'workload-spec' {
  return 'workload-spec'
}

/**
 * interpret validates the profile the way internal/catalog.Validate does for the checks
 * that do not need blis — the name and the group-side knobs — and confirms the spec is
 * parseable YAML, then builds the body to save. An unrunnable form yields a null body
 * (nothing unrunnable is offered); blis's own semantic checks are the server's to add.
 */
export function interpret(v: FormValues): Interpretation {
  const issues: Issue[] = []
  const push = (field: string, message: string) => issues.push({ field, message })

  if (!v.name.trim()) push('name', 'A name is required.')
  else if (!NAME_PATTERN.test(v.name))
    push('name', 'Use lower-case letters, digits, dash, underscore or dot.')
  if (!Number.isInteger(v.seed) || v.seed < 0) push('seed', 'The seed must be a non-negative integer.')
  if (v.requestTimeoutS === 0)
    push(
      'requestTimeoutS',
      '0 is rejected by blis; use a positive deadline, or a negative value to disable it.',
    )
  if (v.horizonTicks !== null && v.horizonTicks <= 0)
    push('horizonTicks', 'The horizon must be greater than 0, or empty for unbounded.')

  const { error } = parseSpec(v.specYaml)
  if (error) push('specYaml', error)

  const workload: WorkloadBody = { type: 'workload-spec', spec_yaml: v.specYaml }
  const body: ProfileBody = {
    name: v.name,
    seed: v.seed,
    horizon_ticks: v.horizonTicks,
    request_timeout_s: v.requestTimeoutS,
    workload,
  }
  return { issues, body: issues.length > 0 ? null : body }
}

/**
 * profileToGroup resolves a profile into the record-shape comparability group, mirroring
 * internal/catalog's resolution so a profile's key lines up with the records already on
 * the board. The group carries no model (E1): model is a candidate supplied per row. A
 * spec profile carries the not-applicable placeholders the record schema uses.
 */
export function profileToGroup(body: ProfileBody): Group {
  const w = body.workload
  let workload: SchemaWorkload
  if (w.type === 'workload-spec') {
    workload = {
      type: 'workload-spec',
      arrival_process: 'constant',
      num_requests: 0,
      load: { kind: 'rate', value: 0 },
      prompt_tokens: 0,
      prompt_tokens_stdev: 0,
      output_tokens: 0,
      output_tokens_stdev: 0,
      spec_file: null,
      spec_sha256: w.spec_sha256 ?? null,
      spec: w.spec,
    } as SchemaWorkload
  } else {
    workload = {
      type: 'distribution',
      arrival_process: w.load!.kind === 'concurrency' ? 'closed-loop' : 'constant',
      num_requests: w.num_requests!,
      load: w.load!,
      prompt_tokens: w.prompt_tokens!,
      prompt_tokens_stdev: w.prompt_tokens_stdev!,
      output_tokens: w.output_tokens!,
      output_tokens_stdev: w.output_tokens_stdev!,
      spec_file: null,
      spec_sha256: null,
    }
  }
  return {
    seed: body.seed,
    horizon_ticks: body.horizon_ticks,
    request_timeout_s: body.request_timeout_s,
    workload,
  }
}

/**
 * profileSummary is the one-line work summary for the catalog browser. A spec profile is
 * summarized by its aggregate rate the way the server's summarize does — never by numbers
 * the spec does not state; a legacy distribution profile reuses workloadTitle so its
 * phrasing cannot drift from the board.
 */
export function profileSummary(body: ProfileBody): string {
  const w = body.workload
  if (w.type === 'workload-spec') {
    const obj = w.spec ? (w.spec as SpecObject) : parseSpec(w.spec_yaml ?? '').obj
    return summarizeSpec(obj)
  }
  return workloadTitle(profileToGroup(body))
}

/**
 * specText is the complete WorkloadSpec of a spec-backed profile, as YAML — the text the
 * catalog's detail panel shows and offers to copy. It prefers the server's own YAML
 * rendering (spec_yaml) and falls back to serializing the spec object, so a profile built
 * client-side (a test, a fresh save) still resolves. A distribution profile has no spec,
 * so this is null and the detail panel shows its flat fields instead.
 */
export function specText(body: ProfileBody): string | null {
  const w = body.workload
  if (w.type !== 'workload-spec') return null
  if (w.spec_yaml && w.spec_yaml.trim()) return w.spec_yaml
  if (w.spec) return serializeSpec(w.spec as SpecObject)
  return null
}

/**
 * profileKnobs is the one-line group-side summary shown above a profile's spec: the seed,
 * the observation window, and the request timeout — the fields that decide whether the
 * declared work can complete, and so are part of the comparability group (CLAUDE.md).
 */
export function profileKnobs(body: ProfileBody): string {
  const horizon = body.horizon_ticks == null ? 'no horizon cap' : `horizon ${body.horizon_ticks} ticks`
  const timeout = body.request_timeout_s < 0 ? 'timeout disabled' : `timeout ${body.request_timeout_s}s`
  return `seed ${body.seed} · ${horizon} · ${timeout}`
}

export interface CrossRef {
  models: number
  runs: number
}

/**
 * crossRef reports how many models and runs already exist on the board under this
 * profile's workload, matched by workloadKey (the topology spec's grouping). Both
 * variants are matched. A spec profile — including the presets — carries the spec and the
 * spec_sha256 the server emits, so profileToGroup rebuilds the same group the runs
 * declared from it were filed under, and its key lines up with theirs. (Before the
 * Declare-a-run flow landed no spec-backed records existed, so this used to short-circuit
 * to zero for a spec profile; that left every preset reading "no runs yet".) A profile the
 * server has not resolved to a spec object yet has no key to match and falls through to
 * zero.
 */
export function crossRef(body: ProfileBody, workloads: WorkloadGroup[]): CrossRef {
  const key = workloadKey(profileToGroup(body))
  const match = workloads.find((w) => w.workloadKey === key)
  return match ? { models: match.models.length, runs: match.records.length } : { models: 0, runs: 0 }
}

// --- API client ------------------------------------------------------------------
// Same shape as newrun.postRun: a fetch rejection is "the server is not running", an
// HTTP error carries the server's {error} message, and fetchImpl is injectable so the
// calls are unit-tested without a server.

const UNREACHABLE =
  'Could not reach the workload server. Start it with `make build && ./bin/leaderboard serve`, then try again.'

async function readOrThrow(res: Response): Promise<unknown> {
  const text = await res.text()
  let parsed: unknown = null
  try {
    parsed = text ? JSON.parse(text) : null
  } catch {
    parsed = null
  }
  if (!res.ok) {
    const message =
      parsed && typeof parsed === 'object' && 'error' in parsed
        ? String((parsed as { error: unknown }).error)
        : `The workload server returned HTTP ${res.status}.`
    throw new Error(message)
  }
  return parsed
}

export async function listWorkloads(fetchImpl: typeof fetch = fetch): Promise<ProfileBody[]> {
  let res: Response
  try {
    res = await fetchImpl('/api/workloads')
  } catch {
    throw new Error(UNREACHABLE)
  }
  const parsed = (await readOrThrow(res)) as { workloads?: ProfileBody[] } | null
  return parsed?.workloads ?? []
}

export async function validateWorkload(
  body: ProfileBody,
  fetchImpl: typeof fetch = fetch,
): Promise<ValidateResponse> {
  let res: Response
  try {
    res = await fetchImpl('/api/workloads/validate', jsonPost('POST', body))
  } catch {
    throw new Error(UNREACHABLE)
  }
  return (await readOrThrow(res)) as ValidateResponse
}

/**
 * saveWorkload creates the profile (POST) when originalName is null, or edits/renames
 * the profile at originalName (PUT) otherwise.
 */
export async function saveWorkload(
  body: ProfileBody,
  originalName: string | null,
  fetchImpl: typeof fetch = fetch,
): Promise<ProfileBody> {
  const url = originalName == null ? '/api/workloads' : `/api/workloads/${encodeURIComponent(originalName)}`
  const method = originalName == null ? 'POST' : 'PUT'
  let res: Response
  try {
    res = await fetchImpl(url, jsonPost(method, body))
  } catch {
    throw new Error(UNREACHABLE)
  }
  return (await readOrThrow(res)) as ProfileBody
}

export async function deleteWorkload(name: string, fetchImpl: typeof fetch = fetch): Promise<void> {
  let res: Response
  try {
    res = await fetchImpl(`/api/workloads/${encodeURIComponent(name)}`, { method: 'DELETE' })
  } catch {
    throw new Error(UNREACHABLE)
  }
  await readOrThrow(res)
}

function jsonPost(method: string, body: unknown): RequestInit {
  return { method, headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(body) }
}
